### PR TITLE
Ruby enumerable methods

### PR DESCRIPTION
**Pull request by alfredoC10**
Greetings, everyone! This project is about rebuilding the Ruby enumerable methods, for this goal I used other methods, mostly iterators and conditionals like if or unless and some methods that were declared before in this same file. Through the code lines, you will find comments that I hope give a comprehensive description of it's use. 
I Can not wait to read and follow your observations, which I really believe will allow a great improved final version.  
Excellent day to you!

### CODE REVIEW

**Comment by alfredoC10** [comment]
I saw that the rubocop linter asked me to make changes, but when doing them, the same rubicop marked those changes as an as errors, for this reason, I disabled some cops like Style/SymbolProc
I'm at your disposal to clarify the details of this action.

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+      'Wrong input'
+    else
+      new_arr = []
+      my_each do |elmt|
+        filtered_arr.push(elmt) if yield(elmt)
+      end
+      new_arr
+    end
+  end
+end
```
Comment (arslanbisharat) [inline enumerables.rb]: * [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_select
- [x] my_select method has two problems. Please have a look below. 

- enumerables #**my_select** returns an array containing all elements of enum for which the given block returns a true value



- enumerable #my_select does not mutate the original array

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    new_arr = []
+
+    if self.class == Array or self.class == Range
+      my_each { |i| new_arr.push(yield(i)) }
+    elsif self.class == Hash
+      my_each { |key, value| new_arr.push(yield(key, value)) }
+    end
+    new_arr
+  end
+end
```
Comment (arslanbisharat) [inline enumerables.rb]: * [x] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [x] returns an Enumerator if no block is given
```ruby
[2,4,7,11].my_map should be equal to [2,4,7,11].map #<Enumerator: [2, 4, 7, 11]:map
```



- [x] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    end
+
+    memo
+  end
+
+  # 10.- ---- Method created for testing my_inject method ----
+
+  def multiply_els(array)
+    array.my_inject(:*)
+  end
```
Comment (arslanbisharat) [inline enumerables.rb]: - [x] enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+      memo = args[0]
+      my_each { |elmt| memo = yield(memo, elmt) }
+
+      # Case 3: only a block is given
+    elsif args.empty? && block_given?
+      self[1..-1].my_each { |elmt| memo = yield(memo, elmt) }
+    end
+
+    memo
+  end
```
Comment (arslanbisharat) [inline enumerables.rb]: - [x] should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [x] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving the wrong result
image

- Real inject
image

- [x] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
 



- Real inject
image

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
# Request changes 😳

I am very happy to see this project and I can see you worked hard on this project. There are only a few issues that we need to address and I am sure that you will end up with a perfect project

_It's okay. We are here to learn. Don't get discouraged. Work even harder. You are almost there.😄_

**Please have a look at the suggestions to get more understanding.**

- [x] For each method you are using a different module. I think that's not necessary. Onlyuse one module for your code. 

**Please make the changes and request re-review for the given project**
Stay safe and give me a smile 😇

### Happy Coding
---------------------------------------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

File: enumerables.rb (hunk 1 above)
Comment (alfredoC10) [reply enumerables.rb]: Good evening Arslan Bisharat! Thanks for the review, I find your words very encouraging and the points you marked, done in time and focused right on the thing to fix! An example is this particular point. Indeed, as you show above, the reason that #my_select wasn't mutating the original array was that I used a variable with a different name than the one defined for this purpose, I've just changed it for the right one, Now, in the example below it really seems that is doing something to the given array:

(comparing #my_select with Ruby's #select) image

(now both outputs) image

All the best!

File: enumerables.rb (hunk 2 above)
Comment (alfredoC10) [reply enumerables.rb]: Arslan Bisharat, hope you are doing great.
Is this a valid output when calling my_map method without any proc or block?

This is what I get when comparing #my_map with #map in Repl.it . I must admit still need to know more about when methods return Enumerators, it took me some time trying to figure out:

image

(IDE output: ) 
image

File: enumerables.rb (hunk 3 above)
Comment (alfredoC10) [reply enumerables.rb]: Thank you for the observation. I had a look at this method, made few changes, and tested on Repl.it. 
This is what is now returning:
image

File: enumerables.rb (hunk 4 above)
Comment (alfredoC10) [reply enumerables.rb]: Understood. I made changes to #my_inject method to work as required and tested in Repl.it. Here the results:

- When looking for #my_inject to return the same as Ruby's #inject 
image

- Raising LocalJumpError
image

- When only a block Is given without initial value
image

- When a symbol is speicified without initial value
image

Im at your attention for any related question

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+      call_block.call
+
+    end
+  end
+
+  # 10.- ---- Method created for testing my_inject method ----
+
+  def multiply_els(array)
+    array.my_inject(:*)
+  end
```
Comment (SpaYco) [inline enumerables.rb]: - [x] `multiply_els` works fine, but it's supposed to be outside the module, take it out and that will be all :)

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

## Required Changes ♻️ 

- [x] check the comment down below 👇 

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Review by ebeagusamuel — APPROVED** [review]
# Project Approved :1st_place_medal: 
Hello @Huemac-Alfredo 
Nice work you have done and also in implementing all the changes required by the previous TSEs'
Your project now passes all the tests and you can go ahead and merge this branch and good luck with the next project
slack(@ebeagusamuel)

**Pull request merged**